### PR TITLE
Add Python typing to all files + Cleanup based off it

### PR DESCRIPTION
**Pull request by abejgonzalez**
I started doing this in my "where to run" PR but figured it should probably be split out since static type checking catches a lot of potential bugs in the code (checking `None`s, improper member variable setup, etc).

This does the following:
- Adds types to all of `runtools` (some `Any` types could be further refined but I'll leave that for a future PR)
- Uses `__future__.annotations` everywhere (to make annotations strings - speeds things up) - Fixes bugs based on this in the `register_task` `firesim` decorator
- Avoids the use of `from ... import *` - Needed so that global variables aren't redefined (causing the type checker to not infer certain variables like `rootLogger`)
- Reorders the `__init__` of `FireSimTopology` to the top of the class (like other classes we write)
- Uses `f"..."` for the `basecommand` in `RuntimeConfig`
- Reworks the `add_simulation` and `add_switch` of the run farm instances (in `run_farm.py`) to not use `None`s. This should be the same functionality since the `None`s were filled in from 0 to MAX (i.e. slots were filled contiguously). Do we care about the other potential functionality of having the slots filled in a non-contiguous manner (slot 0, 3, 5, 6 filled)? IMO I don't think we care about this on F1, and in the future we can just override this on different run farm host types.

This depends on #1037 going in.

Sorry in advance for the big PR (much of it is pretty easy to understand).

**TODO**:
- [x] Fix `rootLogger` type inference
- [x] Misc. cleanup
- [x] Make sure there are no circular dependencies
- [ ] Run FPGA sim to confirm all is working (will run when this is about to be merged into main through CI)

### Contributor Checklist
- [x] Is this PR's title suitable for inclusion in the changelog and have you added a `changelog:<topic>` label?
- [x] Did you add Scaladoc/docstring/doxygen to every public function/method?
- [x] Did you add at least one test demonstrating the PR?
- [x] Did you delete any extraneous prints/debugging code?
- [x] Did you state the UI / API impact?
- [x] Did you specify the Verilog / AGFI compatibility impact?

- [x] If applicable, did you regenerate and publicly share default AGFIs?

- [x] If applicable, did you apply the `ci:fpga-deploy` label?

- [x] If applicable, did you apply the `Please Backport` label?

### Reviewer Checklist (only modified by reviewer)
- [ ] Is the title suitable for inclusion in the changelog and does the PR have a `changelog:<topic>` label?
- [ ] Did you mark the proper release milestone?
- [ ] Did you check whether all relevant Contributor checkboxes have been checked?

### CODE REVIEW

File: deploy/runtools/firesim_topology_elements.py
Comment (abejgonzalez) [inline deploy/runtools/firesim_topology_elements.py:588]: General comment: Across all of this code I have to add these asserts so that the type checker doesn't error. These need to be analyzed much more to make sure they are in the right places (or maybe we can remove some extra `Optional` types).

File: deploy/runtools/utils.py
Comment (abejgonzalez) [inline deploy/runtools/utils.py:368]: @timsnyder-siv the type checker errors here since dso.path could potentially be `None`. What should this be?

File: deploy/runtools/firesim_topology_elements.py
Comment (abejgonzalez) [inline deploy/runtools/firesim_topology_elements.py:291]: Probably can create a new function called something like `resolved_server_hw_cfg` that has these asserts built-in. Maybe we can have the `server_hw_cfg` be split into `server_hw_cfg` and `server_hw_cfg_name` so that we don't need this?

File: deploy/runtools/firesim_topology_elements.py
Comment (abejgonzalez) [inline deploy/runtools/firesim_topology_elements.py:397]: IDK why but this was surprisingly hard to get `mypy` to cooperate hence why this is verbose now.

File: deploy/runtools/firesim_topology_elements.py
Comment (abejgonzalez) [inline deploy/runtools/firesim_topology_elements.py:452]: Figured this was too verbose? I just inlined these specific calls further down the file. I can revert this if people want.

File: deploy/runtools/firesim_topology_elements.py
Comment (abejgonzalez) [inline deploy/runtools/firesim_topology_elements.py:534]: Can probably convert this back into the `[self.supernode_get_sibling(x).FOO_TO_CALL]` syntax

File: deploy/runtools/firesim_topology_elements.py
Comment (abejgonzalez) [inline deploy/runtools/firesim_topology_elements.py:563]: Who tests supernode anyways?!?

File: deploy/runtools/firesim_topology_with_passes.py
Comment (abejgonzalez) [inline deploy/runtools/firesim_topology_with_passes.py:263]: Again, this is more verbose because `mypy` can't detect that `alldownlinknodes` are of type `FireSimServerNode` from this `if` statement. Maybe this is re-workable but this works for now.

File: deploy/runtools/firesim_topology_with_passes.py
Comment (abejgonzalez) [inline deploy/runtools/firesim_topology_with_passes.py:317]: I personally think this code is very confusing... in some places you expected the HW_CFG to be a `str` and in others you wanted a `RuntimeHWConfig` object. This keeps the same functionality but IMO this is ugly.

File: deploy/runtools/run_farm.py
Comment (abejgonzalez) [inline deploy/runtools/run_farm.py:138]: I reworked the `add_simulation` and `add_switch` to avoid having to use `Optional` types. This is still the same functionality as before.

File: deploy/runtools/user_topology.py
Comment (abejgonzalez) [inline deploy/runtools/user_topology.py:266]: This errored since a `List[SuperNode]` couldn't be added to a `List[DummyNode]`s.

File: deploy/runtools/user_topology.py
Comment (abejgonzalez) [inline deploy/runtools/user_topology.py:4]: Having `import *` statements in the code caused things like the `rootLogger` global variable to be imported from other modules (and thus making mypy unable to infer it's type). Across the codebase I removed this and explicitly imported each object/class.

File: deploy/runtools/runtime_config.py
Comment (abejgonzalez) [inline deploy/runtools/runtime_config.py:142]: Renamed this so that the `basecommand` line could be reduced to a `f"..."` string.

File: deploy/runtools/runtime_config.py
Comment (abejgonzalez) [inline deploy/runtools/runtime_config.py:24]: Unused code.

File: deploy/runtools/run_farm.py
Comment (abejgonzalez) [inline deploy/runtools/run_farm.py:104]: See https://github.com/firesim/firesim/pull/1042/files#r857293207

File: deploy/runtools/run_farm.py
Comment (abejgonzalez) [inline deploy/runtools/run_farm.py:396]: This is hacky because lists of different types can't be added together. The type checker doesn't work in this scenario: 

```
class A(X)
class B(X)
class C(X)

all: List[X] = [A] + [B] + [C] # errors since A can't be added to B (doesn't resolve to X)
```

File: deploy/runtools/firesim_topology_elements.py
Comment (abejgonzalez) [reply deploy/runtools/firesim_topology_elements.py]: Ok I did this.

File: deploy/runtools/firesim_topology_elements.py
Comment (abejgonzalez) [inline deploy/runtools/firesim_topology_elements.py:227]:
```suggestion
        assert self.server_hardware_config is not None and not isinstance(self.server_hardware_config, str)
```

File: deploy/runtools/utils.py
Comment (timsnyder-siv) [reply deploy/runtools/utils.py:370]: When I checkout your `runtools-mypy` branch and `./scripts/run-py-typecheck.sh`, I don't see an error.  What am I missing?

File: deploy/awstools/afitools.py
Comment (timsnyder-siv) [inline deploy/awstools/afitools.py:3]: Thank you for doing this across all files.  Since it changes the fundamental behavior of type annotations, it is good to be consistent in how we're using it.  It would be nice if mypy had a check (or something somewhere) that makes sure that we continue to have the import in all python files.

File: deploy/runtools/user_topology.py
Comment (timsnyder-siv) [reply deploy/runtools/user_topology.py:266]: Even if you do this?
```
servers : List[FireSimServerNode] = [FireSimSuperNodeServerNode()] + [FireSimDummyServerNode() for x in range(5)]
```

from https://mypy.readthedocs.io/en/latest/common_issues.html#variance

Comment (timsnyder-siv) [reply deploy/runtools/user_topology.py:266]: For my own edification, I confirmed that with this:
```
diff --git a/deploy/runtools/user_topology.py b/deploy/runtools/user_topology.py
index 4336b4b14..a7173fd37 100644
--- a/deploy/runtools/user_topology.py
+++ b/deploy/runtools/user_topology.py
@@ -5,7 +5,7 @@ from __future__ import annotations
 
 from runtools.firesim_topology_elements import FireSimSwitchNode, FireSimServerNode, FireSimSuperNodeServerNode, FireSimDummyServerNode, FireSimNode
 
-from typing import Optional, Union, Callable, Sequence, TYPE_CHECKING
+from typing import List, Optional, Union, Callable, Sequence, TYPE_CHECKING
 if TYPE_CHECKING:
     from runtools.firesim_topology_with_passes import FireSimTopologyWithPasses
 
@@ -275,8 +275,8 @@ class UserTopologies:
 
     def supernode_example_6config(self) -> None:
         self.roots = [FireSimSwitchNode()]
-        self.roots[0].add_downlinks([FireSimSuperNodeServerNode()])
-        self.roots[0].add_downlinks([FireSimDummyServerNode() for x in range(5)])
+        servers : List[FireSimServerNode] = [FireSimSuperNodeServerNode()] + [FireSimDummyServerNode() for x in range(5)]
+        self.roots[0].add_downlinks(servers)
 
     def supernode_example_4config(self) -> None:
         self.roots = [FireSimSwitchNode()]
Submodule target-design/chipyard contains modified content
```

I still see:
```
\u276f ./scripts/run-py-typecheck.sh 
deploy/runtools/user_topology.py:278: error: Incompatible types in assignment (expression has type "List[FireSimSuperNodeServerNode]", variable has type "List[FireSimServerNode]")
deploy/runtools/user_topology.py:278: note: "List" is invariant -- see https://mypy.readthedocs.io/en/stable/common_issues.html#variance
deploy/runtools/user_topology.py:278: note: Consider using "Sequence" instead, which is covariant
deploy/runtools/user_topology.py:278: error: List comprehension has incompatible type List[FireSimDummyServerNode]; expected List[FireSimSuperNodeServerNode]
Found 2 errors in 1 file (checked 22 source files)

```

Comment (timsnyder-siv) [reply deploy/runtools/user_topology.py:266]: This works:
```
diff --git a/deploy/runtools/user_topology.py b/deploy/runtools/user_topology.py
index 4336b4b14..ce3479c1b 100644
--- a/deploy/runtools/user_topology.py
+++ b/deploy/runtools/user_topology.py
@@ -5,7 +5,7 @@ from __future__ import annotations
 
 from runtools.firesim_topology_elements import FireSimSwitchNode, FireSimServerNode, FireSimSuperNodeServerNode, FireSimDummyServerNode, FireSimNode
 
-from typing import Optional, Union, Callable, Sequence, TYPE_CHECKING
+from typing import cast, List, Optional, Union, Callable, Sequence, TYPE_CHECKING
 if TYPE_CHECKING:
     from runtools.firesim_topology_with_passes import FireSimTopologyWithPasses
 
@@ -275,8 +275,8 @@ class UserTopologies:
 
     def supernode_example_6config(self) -> None:
         self.roots = [FireSimSwitchNode()]
-        self.roots[0].add_downlinks([FireSimSuperNodeServerNode()])
-        self.roots[0].add_downlinks([FireSimDummyServerNode() for x in range(5)])
+        servers = [cast(FireSimServerNode, FireSimSuperNodeServerNode())] + [cast(FireSimServerNode, FireSimDummyServerNode()) for x in range(5)]
+        self.roots[0].add_downlinks(servers)
 
     def supernode_example_4config(self) -> None:
         self.roots = [FireSimSwitchNode()]
Submodule target-design/chipyard contains modified content

```

But I like the way you rewrote it to avoid the problem better.  I really hate having to rewrite reasonable code for static checkers but whatever.

**Comment by abejgonzalez** [comment]
@timsnyder-siv IDK why I can't comment on this: https://github.com/firesim/firesim/pull/1042#discussion_r859030911. But I added the extra `if dso.path is not None:` check. Is that correct?

File: deploy/runtools/utils.py
Comment (timsnyder-siv) [reply deploy/runtools/utils.py:370]: > IDK why I can't comment on this: https://github.com/firesim/firesim/pull/1042#discussion_r859030911. But I added the extra if dso.path is not None: check. Is that correct?

I'm only able to comment in the thread if I go to the "Files Changed" tab, not in Conversations for some reason.

I looked back though lddwrap and I think I really meant 
```
       assert dso.path is not None and '/ld-linux' in fspath(dso.path), ...
```

I don't want it to silently continue past anything other than the dynamic linker in that case.

File: deploy/runtools/utils.py
Comment (timsnyder-siv) [inline deploy/runtools/utils.py:371]:
```suggestion
            assert dso.path is not None and '/ld-linux' in fspath(dso.path), f"dynamic linker is only allowed no soname, not: {dso}"
```

**Review by timsnyder-siv — CHANGES_REQUESTED** [review]
Definitely should change this before merging.

File: deploy/runtools/runtime_config.py
Comment (abejgonzalez) [reply deploy/runtools/runtime_config.py:149]: I did this in the #1028 so I'll skip it in this PR for now.

File: deploy/runtools/run_farm.py
Comment (abejgonzalez) [inline deploy/runtools/run_farm.py:390]: This should be safe because you assign IP addresses instances before this call.

File: deploy/runtools/firesim_topology_with_passes.py
Comment (abejgonzalez) [inline deploy/runtools/firesim_topology_with_passes.py:580]: This is fine because it is assigned in the top-level `firesim` py script.

File: deploy/runtools/run_farm.py
Comment (timsnyder-siv) [inline deploy/runtools/run_farm.py:522]: How does this look once you have 'what-to-run' implemented?  I strongly dislike the current code silently does nothing `unless isinstance(self.parentnode, F1_Instance)` if for no other reason than we no longer "always clear ALL fpga slots" when `clear_fpgas` is called.  Should probably make this a `match` statement with default case that throws `NotImplemented` so that we don't open the door for really difficult to understand silent failure cases.

However, if 'where-to-run' takes care of this by abstracting this method behind a dispatcher that has to implement this method, I'm willing to look the other way here.

File: deploy/runtools/firesim_topology_elements.py
Comment (sagark) [inline deploy/runtools/firesim_topology_elements.py:667]:
```suggestion
        msg =  f"FireSimSwitchNode:{self.switch_id_internal}\n"
```

File: deploy/runtools/firesim_topology_with_passes.py
Comment (sagark) [inline deploy/runtools/firesim_topology_with_passes.py:306]:
```suggestion
            it with the appropriate RuntimeHWConfig object. If it is already a
```

File: deploy/runtools/run_farm.py
Comment (abejgonzalez) [reply deploy/runtools/run_farm.py:522]: These if's are really no-ops (with how things are currently used). In reality, these functions are called properly in `infrasetup` (https://github.com/firesim/firesim/blob/043905e14ddb304705226738df14dc8885faf436/deploy/runtools/run_farm.py#L676)(or other types of functions) with an `if` in that function doing the same thing. If the preference is adding an assert instead of silently skipping that's fine with me. I was just trying to avoid adding typing changes that affect the original code too much.

**Review by sagark — APPROVED** [review]
LGTM modulo small suggestions.

File: deploy/runtools/run_farm.py
Comment (abejgonzalez) [reply deploy/runtools/run_farm.py:522]: @timsnyder-siv is this the only issue with the PR? If so I'll address this and I think we can merge this (since it doesn't change any functionality)

**Comment by abejgonzalez** [comment]
@mergifyio rebase

File: deploy/runtools/run_farm.py
Comment (abejgonzalez) [reply deploy/runtools/run_farm.py]: Solved offline. We will revert to using `assert`s

**Pull request merged**
